File: windows/src/TraceCommons.Interop/HealthCopy.cs
using System;

namespace TraceCommons.Interop;

/// <summary>
/// The health banner's words, transcribed from the shared design's
/// failure-state table (2026-08-08, "Failure states") and split into the
/// title/body pair the visual design draws for the Windows frame
/// (2026-08-17 section 5.1 item 3).
///
/// Two rules bind every sentence here, and both are the design's:
/// <list type="bullet">
/// <item><b>Never name the mechanism.</b> "Privacy filter", "claim",
/// "ingest", "canary" and "PII" are internal words. The label carries them;
/// the sentence must not.</item>
/// <item><b>Always state the data consequence.</b> "Nothing has been lost",
/// "your queue is safe", "rather than going out unscanned".</item>
/// </list>
/// </summary>
/// <remarks>
/// <para>
/// This is a pure one-label-to-one-banner mapping and it must stay that way.
/// <c>status.health.last_error_label</c> carries a single label that the
/// daemon has already resolved through its own precedence order
/// (<c>daemon::health::precedence</c>: not-logged-in outranks the near-AI
/// notice, which outranks the self-test failure, and so on down). A client
/// that reconstructed that order would eventually disagree with the daemon
/// about what is wrong, and a contributor would be told one thing by the tray
/// and a different thing by the window. So this renders whichever label
/// arrives and never ranks, merges, or synthesises one. The Linux shell's
/// <c>render_health</c> carries the same note for the same reason.
/// </para>
/// <para>
/// It lives in the interop assembly rather than in a view model for the same
/// reason <see cref="ReadGate"/> does: it is the wording half of a safety
/// surface, and here it is exercised by tests on a machine that cannot build
/// WinUI at all.
/// </para>
/// </remarks>
public sealed class HealthCopy : IEquatable<HealthCopy>
{
    private HealthCopy(string title, string detail, string? actionLabel)
    {
        Title = title;
        Detail = detail;
        ActionLabel = actionLabel;
    }

    /// <summary>The banner's first line, 13px/600 in the frame.</summary>
    public string Title { get; }

    /// <summary>The rest of the sentence, and where the consequence is stated.</summary>
    public string Detail { get; }

    /// <summary>
    /// The banner's action, for conditions with a contributor recovery step.
    /// </summary>
    /// <remarks>
    /// Null for everything else, and deliberately so. The other conditions
    /// clear on their own, and a button that cannot change the condition it
    /// sits beside teaches a contributor that the buttons in this app do
    /// nothing -- which is a lesson they would then apply to Undo.
    /// </remarks>
    public string? ActionLabel { get; }

    /// <summary>
    /// The banner for a health label, or null when there is nothing wrong.
    /// </summary>
    /// <remarks>
    /// A null or empty label is health, not an unknown condition: the daemon
    /// expresses "fine" as the absence of a label.
    /// </remarks>
    public static HealthCopy? ForLabel(string? label)
    {
        if (string.IsNullOrEmpty(label))
        {
            return null;
        }

        return label switch
        {
            "not-logged-in" => new HealthCopy(
                "Not connected.",
                "Sessions are being queued, but nothing can be sent until you reconnect. "
                + "Nothing has been lost.",
                "Reconnect"),
            "near-ai-notice-not-acknowledged" => new HealthCopy(
                "One thing to confirm.",
                "You chose the extra privacy scan, which sends message text to NEAR AI. "
                + "Confirm you're OK with that and contributions resume.",
                "Review and confirm"),
            "privacy-filter-canary-failed" => new HealthCopy(
                "The privacy scan failed its own self-test,",
                "so nothing is being sent through it. This is deliberate -- a scan we can't "
                + "verify doesn't get used.",
                null),
            "pii-filter-unavailable" => new HealthCopy(
                "The extra privacy scan isn't reachable.",
                "Your traces are waiting rather than going out unscanned. Retrying "
                + "automatically.",
                null),
            "claim-mint-failed" or "ingest-unreachable" => new HealthCopy(
                "Can't reach Trace Commons right now.",
                "Your queue is safe; it'll retry on its own.",
                null),
            "queue-full" => new HealthCopy(
                "Trace Commons has stopped queuing new sessions",
                "-- 500 are already waiting. Review or clear some to start again.",
                "Review"),
            // The fallback for a daemon that reported the label but no
            // daily_budget object. ForBudget is what normally renders this
            // condition, because it can say how many are waiting and when
            // the limit actually resets. This line used to promise "the
            // rest goes out tomorrow", which the daemon never said: it
            // rolls its counters at UTC midnight, which is not tomorrow for
            // most of the world.
            "daily-cap-reached" => new HealthCopy(
                BudgetTitle,
                "Approved traces are waiting. Nothing has been lost -- they go out when the "
                + "limit resets.",
                null),

            // An unrecognised label is still a real condition, and the daemon
            // is free to grow labels this build has never heard of. Say the
            // thing that holds for every blocking label rather than inventing
            // a cause, and never render the raw label as the explanation: a
            // label is an internal name, and showing it would break the
            // never-name-the-mechanism rule by the most direct route there is.
            _ => new HealthCopy(
                "Contributions are on hold.",
                "Something is stopping traces from being sent. Nothing has been lost, and "
                + "nothing has gone out.",
                null),
        };
    }

    /// <summary>The banner title for a spent daily budget.</summary>
    public const string BudgetTitle = "Today's upload limit is used up.";

    /// <summary>
    /// The banner for a spent daily budget, built from the numbers the
    /// daemon reported, or null when nothing is being held back.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Deliberately not part of <see cref="ForLabel"/>, and deliberately not
    /// subject to the daemon's health precedence. <c>daily-cap-reached</c>
    /// is last in that order, so on the machine this was written for the
    /// single health slot was held by <c>queue-full</c> and the real reason
    /// approvals were not moving never reached a screen. A window that waits
    /// for the label will keep missing the condition; it must read
    /// <c>status.daily_budget</c> instead.
    /// </para>
    /// <para>
    /// No action label: there is nothing a contributor can do about it, and
    /// the caps are not settable from here.
    /// </para>
    /// </remarks>
    public static HealthCopy? ForBudget(DailyBudget? budget)
    {
        if (budget is null || !budget.Blocked)
        {
            return null;
        }

        var waiting = budget.BlockedEntries switch
        {
            <= 0 => "Approved traces are waiting",
            1 => "1 approved trace is waiting",
            var n => $"{n} approved traces are waiting",
        };

        var resets = budget.ResetsAtUtc;
        var detail = resets is null
            ? $"{waiting}. Nothing has been lost -- they go out when the limit resets."
            : $"{waiting}. Nothing has been lost -- they go out when the limit resets at "
                + $"{resets.Value.ToLocalTime():t}.";

        return new HealthCopy(BudgetTitle, detail, null);
    }

    public bool Equals(HealthCopy? other) =>
        other is not null
        && Title == other.Title
        && Detail == other.Detail
        && ActionLabel == other.ActionLabel;

    public override bool Equals(object? obj) => Equals(obj as HealthCopy);

    public override int GetHashCode() => HashCode.Combine(Title, Detail, ActionLabel);
}

File: macos/Sources/TraceCommonsApp/HealthCopy.swift
import Foundation
import TCShellCore

/// The health sentences, verbatim from the shared design's failure-state
/// table. Two rules hold across every one of them: never name the mechanism
/// ("privacy filter", "claim", "ingest", "canary" are internal words), and
/// always state the data consequence.
///
/// `status.health.last_error_label` carries ONE label at a time, already
/// resolved by the daemon's precedence order. This table does not
/// reconstruct that order and must not: it renders whichever label arrives.
struct HealthCopy: Equatable {
    enum Severity {
        /// Something the contributor can act on.
        case actionable
        /// Ambient; it clears on its own.
        case waiting
        /// Menu line only.
        case informational
    }

    let title: String
    let detail: String
    let severity: Severity
    /// Present only where there is a real action behind it.
    let actionTitle: String?
    var reviewsQueue = false

    /// The banner for a spent daily budget, built from the numbers the
    /// daemon actually reported.
    ///
    /// Separate from `forLabel` because the daemon reports this separately:
    /// `daily-cap-reached` is last in the precedence order, so on the
    /// machine this was written for the single health slot was held by
    /// `queue-full` and the real reason nothing was uploading never reached
    /// a screen. A shell that waits for the label to arrive will keep
    /// missing it.
    ///
    /// `.waiting`, not an error: nothing is broken and nothing was lost.
    static func forBudget(_ budget: DailyBudget) -> HealthCopy? {
        guard budget.blocked else { return nil }
        return HealthCopy(
            title: DailyBudgetCopy.title,
            detail: DailyBudgetCopy.detail(
                blockedEntries: budget.blockedEntries,
                resetsAt: budget.resetsAt
            ),
            severity: .waiting,
            actionTitle: nil
        )
    }

    static func forLabel(_ label: String) -> HealthCopy {
        switch label {
        case "not-logged-in":
            return HealthCopy(
                title: "Not connected.",
                detail: """
                Sessions are being queued, but nothing can be sent until you \
                reconnect. Nothing has been lost.
                """,
                severity: .actionable,
                actionTitle: "Reconnect"
            )
        case "near-ai-notice-not-acknowledged":
            return HealthCopy(
                title: "One thing to confirm.",
                detail: """
                You chose the extra privacy scan, which sends message text to \
                NEAR AI. Confirm you're OK with that and contributions resume.
                """,
                severity: .actionable,
                actionTitle: "Review and confirm"
            )
        case "privacy-filter-canary-failed":
            return HealthCopy(
                title: "The privacy scan failed its own self-test,",
                detail: """
                so nothing is being sent through it. This is deliberate -- a scan \
                we can't verify doesn't get used.
                """,
                severity: .waiting,
                actionTitle: nil
            )
        case "pii-filter-unavailable":
            return HealthCopy(
                title: "The extra privacy scan isn't reachable.",
                detail: """
                Your traces are waiting rather than going out unscanned. Retrying \
                automatically.
                """,
                severity: .waiting,
                actionTitle: nil
            )
        case "claim-mint-failed", "ingest-unreachable":
            return HealthCopy(
                title: "Can't reach Trace Commons right now.",
                detail: "Your queue is safe; it'll retry on its own.",
                severity: .waiting,
                actionTitle: nil
            )
        case "queue-full":
            return HealthCopy(
                title: "Trace Commons has stopped queuing new sessions",
                detail: """
                -- 500 are already waiting. Review or clear some to start again.
                """,
                severity: .actionable,
                actionTitle: "Review",
                reviewsQueue: true
            )
        case "daily-cap-reached":
            // The fallback for a daemon that reported the label without a
            // `daily_budget` object. `forBudget` is what normally renders
            // this condition, and it can say how many are waiting and when
            // the limit actually resets; this line must not promise a time
            // it does not have. It said "The rest goes out tomorrow",
            // which is false for most of the world -- the daemon rolls its
            // counters at UTC midnight.
            return HealthCopy(
                title: DailyBudgetCopy.title,
                detail: """
                Approved traces are waiting. Nothing has been lost -- they go out when the \
                limit resets.
                """,
                severity: .waiting,
                actionTitle: nil
            )
        default:
            // An unrecognised label is still a real condition. Say that
            // something is holding contributions rather than inventing a
            // cause, and never render the raw label as an explanation.
            return HealthCopy(
                title: "Contributions are on hold.",
                detail: """
                Something is stopping traces from being sent. Nothing has been \
                lost, and nothing has gone out.
                """,
                severity: .waiting,
                actionTitle: nil
            )
        }
    }
}

/// Plain-English names for the queue states a contributor sees. Four of them
/// mean nothing left the machine, and each says so in words.
enum QueueStateCopy {
    static func sentence(for state: QueueState) -> String {
        switch state {
        case .pending: return "Waiting for your decision. Nothing has been sent."
        case .approved: return "You said yes. Not sent yet."
        case .uploading: return "Being sent now."
        case .uploaded: return "In the commons."
        case .refused: return "The system declined to send this. Nothing was sent."
        case .failed: return "Sending didn't work. Nothing was sent; it will retry."
        case .expired: return "Dropped after waiting too long for a decision. Never sent."
        case .superseded:
            return """
            This session changed after you approved it, so it was not sent. A \
            fresh copy is waiting for a new decision.
            """
        }
    }
}

/// Plain-English reasons for `queue_outcome_counts`. It covers entries that
/// ARE on the queue -- it cannot explain a session the watcher discarded
/// before an entry existed, and this UI does not claim otherwise.
enum OutcomeCopy {
    static func sentence(for label: String) -> String {
        switch label {
        case "dismissed-by-contributor": return "You said no thanks"
        case "expired-without-decision": return "Waited too long without a decision"
        case "session-changed-after-offer": return "Changed after it was offered"
        case "not-logged-in": return "Waiting until you reconnect"
        case "daily-cap-reached": return "Waiting for tomorrow's allowance"
        case "queue-full": return "Queue was full"
        case "ingest-unreachable", "claim-mint-failed": return "Trace Commons was unreachable"
        case "pii-filter-unavailable": return "Waiting for the extra privacy scan"
        case "privacy-filter-canary-failed": return "The privacy scan failed its self-test"
        default: return "Held"
        }
    }
}
